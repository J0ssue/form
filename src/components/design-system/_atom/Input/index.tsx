import React from "react";
import {
  InputContainer,
  InputComponent,
  InputLabel,
} from "./styled-components";

interface Props {
  type: string;
  label: string;
  placeholder: string;
  hasError: boolean;
  onChange?: () => void;
}

const Input = (props: Props) => {
  const { type, hasError, label, placeholder, onChange } = props;
  return (
    <InputContainer>
      <InputLabel hasError={hasError}>{label}</InputLabel>
      <InputComponent
        type={type}
        onChange={onChange}
        placeholder={placeholder}
        hasError={hasError}
      ></InputComponent>
    </InputContainer>
  );
};

Input.defaultProps = {
  label: "input",
  type: "text",
  placeholder: "string",
  hasError: false,
};

export default Input;
