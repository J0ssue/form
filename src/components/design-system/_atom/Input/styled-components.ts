import styled from "@emotion/styled";
import { css } from "@emotion/react";
import { buildColor } from "components/design-system/_static/colorPalette";

export const InputContainer = styled.div`
  width: 309px;
  height: 48px;
  position: relative;

  & * {
    box-sizing: border-box;
  }
`;

export const InputComponent = styled.input<{ hasError: undefined | string }>`
  width: 100%;
  height: 100%;
  padding: 0 14px 0 14px;
  border-radius: 8px;
  border: 1.5px solid #cfcfcf;
  caret-color: ${buildColor("red", "100")};
  font-size: 14px;
  outline: none;
  transition: border 0.2s, outline 0.2s;

  &:focus {
    border: 1px solid ${buildColor("red", "100")};
  }

  ${({ hasError }) =>
    hasError &&
    css`
      border: 1.5px solid ${buildColor("red", "100")};
      &:focus {
        border: 1.5px solid ${buildColor("red", "100")};
      }
    `}
`;

export const InputTextContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: absolute;
  top: -19px;
  left: 0px;
`;

export const InputLabel = styled.span<{ hasError: Function }>`
  font-size: 12px;
  color: ${({ hasError }) =>
    hasError() ? buildColor("red", "100") : buildColor("black", "900")};
`;

export const InputError = styled.span`
  text-align: right;
  font-size: 12px;
  color: ${buildColor("red", "100")};
`;
