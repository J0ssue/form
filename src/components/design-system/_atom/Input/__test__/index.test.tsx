import Input from "..";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";

describe("Input component renering and behavior", () => {
  const error = {
    message: "error",
  };

  it("should render <Input/> correctly", () => {
    const tree = renderer
      .create(
        <Input type="text" label="Text Input" placeholder="write something" />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render <Input/> with error", () => {
    const tree = renderer
      .create(
        <Input
          type="text"
          label="Text Input Error"
          placeholder="write something"
          error={error}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("<Input/> should display correct text when typed into", () => {
    render(
      <Input
        type="text"
        label="Text Input Typed"
        placeholder="write something"
        error={error}
      />
    );
    userEvent.type(screen.getByRole("textbox"), "Hello, World!");
    expect(screen.getByRole("textbox")).toHaveValue("Hello, World!");
  });
});
