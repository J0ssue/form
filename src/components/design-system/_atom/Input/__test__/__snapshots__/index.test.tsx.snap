// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component renering and behavior should render <Input/> correctly 1`] = `
<div
  className="input-container css-7ya7e0"
>
  <label
    className="css-savvye"
  >
    Text Input
  </label>
  <input
    className="css-1pqnz1a"
    placeholder="write something"
    type="text"
  />
</div>
`;

exports[`Input component renering and behavior should render <Input/> with error 1`] = `
<div
  className="input-container css-7ya7e0"
>
  <label
    className="css-uvk7o0"
  >
    Text Input Error
  </label>
  <input
    className="css-noagsx"
    placeholder="write something"
    type="text"
  />
</div>
`;
